_posts/2023-11-15-Codesgiving-Part-1.md

```diff
@@ -0,0 +1,289 @@
+If there aren't yet any code contributions and it doesn't look like anyone is working on one, consider volunteering to take a further look into the solution and be sure to acknowledge the existing discussions.
+If you're unsure, it's always kind to mention your interest in the report and ask for more information.
+
+### Is the Need a Bug or Feature?
+
+```mermaid!
+flowchart LR
+
+subgraph software_needs ["Software Needs"]
+subgraph bug_over ["Bug"]
+    bug["Something's\nbroken!"]
+end
+
+subgraph feature_over ["Feature"]
+    feature["We need new\nfunctionality!"]
+end
+end
+
+style software_needs fill:#fff,stroke:#fff;
+```
+
+One way to help solidify your thinking and the approach is to consider whether what you're proposing is a bug or a feature.
+A [software bug](https://en.wikipedia.org/wiki/Software_bug) is considered something which is broken or malfunctioning.
+A [software feature](https://en.wikipedia.org/wiki/Software_feature) is generally considered new functionality or a different way of doing things than what exists today.
+There's often overlap between these, and sometimes they can inspire branching needs, but individually they usually are more of one than the other.
+If you can't decide whether your need is a bug or a feature, consider breaking it down into smaller sub-components so they can be more of one or the other.
+Following this strategy will help you communicate the potential for contribution and also clarify the development process.
+
+### Reporting the Need for Change
+
+```markdown
+# Using `function_x` with `library_y` causes `exception_z`
+
+## Summary
+
+As a `library_y` research software developer I want to use `function_x` for my data so that I can share data for research outcomes.
+
+## Reproducing the error
+
+This error may be seen using Python v3.x on all major OS's using the following code snippet:
+...
+
+```
+
+_An example of a user story issue report with imagined code example._
+
+Open-source needs are often best reported through written stories captured within a bug or feature tracking system (such as [GitHub Issues](https://github.com/features/issues)) which if possible also include example code or logs.
+One template for reporting issues is through a "user story".
+A user story typically comes in the form: `As a < type of user >, I want < some goal > so that < some reason >.` ([Mountain Goat Software: User Stories](https://www.mountaingoatsoftware.com/agile/user-stories)).
+Alongside the story, it can help to add in a snippet of code which exemplifies a problem, new functionality, or a potential adjacent / similar solution.
+As a general principle, be as __be as specific as you can without going overboard__.
+Include things like programming language version, operating system, and other system dependencies that might be related.
+Once you have a good written description of the need be sure to submit it where it can be seen by the relevant development community.
```

Comment:
```suggestion
Once you have a good written description of the need, be sure to submit it where it can be seen by the relevant development community.
```

Reply:
Great catch!